be/src/runtime/sender_queue.cpp

```diff
@@ -0,0 +1,957 @@
+    if (_is_cancelled) {
+        return Status::Cancelled("Cancelled SenderQueueForPipeline::get_chunk");
+    }
+    size_t index = _is_pipeline_level_shuffle ? driver_sequence : 0;
+    auto& chunk_queue = _chunk_queues[index];
+
+    ChunkItem item;
+    if (!chunk_queue.try_dequeue(item)) {
+        _unpluging[index] = false;
+        VLOG_ROW << "DataStreamRecvr no new data, stop unpluging";
+        return Status::OK();
+    }
+    if (item.chunk_ptr == nullptr) {
+        ChunkUniquePtr chunk_ptr = std::make_unique<vectorized::Chunk>();
+        faststring uncompressed_buffer;
+        RETURN_IF_ERROR(_deserialize_chunk(item.pchunk, chunk_ptr.get(), &uncompressed_buffer));
+        *chunk = chunk_ptr.release();
+    } else {
+        *chunk = item.chunk_ptr.release();
+    }
+    auto* closure = item.closure;
+    VLOG_ROW << "DataStreamRecvr fetched #rows=" << (*chunk)->num_rows();
+    if (closure != nullptr) {
+#ifndef BE_TEST
+        MemTracker* prev_tracker = tls_thread_status.set_mem_tracker(ExecEnv::GetInstance()->process_mem_tracker());
+        DeferOp op([&] { tls_thread_status.set_mem_tracker(prev_tracker); });
+#endif
+        _recvr->_closure_block_timer->update(MonotonicNanos() - item.queue_enter_time);
+        closure->Run();
+        _blocked_closure_num[index]--;
+    }
+    _total_chunks--;
+    _recvr->_num_buffered_bytes -= item.chunk_bytes;
+    return Status::OK();
+}
+
+bool DataStreamRecvr::PipelineSenderQueue::has_chunk() {
+    if (_is_cancelled) {
+        return true;
+    }
+    if (_chunk_queues[0].size_approx() == 0 && _num_remaining_senders > 0) {
+        return false;
+    }
+    return true;
+}
+
+bool DataStreamRecvr::PipelineSenderQueue::try_get_chunk(vectorized::Chunk** chunk) {
+    if (_is_cancelled) {
+        return false;
+    }
+    auto& chunk_queue = _chunk_queues[0];
+    ChunkItem item;
+    if (!chunk_queue.try_dequeue(item)) {
+        return false;
+    }
+    CHECK(item.chunk_ptr != nullptr);
+    *chunk = item.chunk_ptr.release();
+    VLOG_ROW << "DataStreamRecvr fetched #rows=" << (*chunk)->num_rows();
+    auto* closure = item.closure;
+    if (closure != nullptr) {
+        _recvr->_closure_block_timer->update(MonotonicNanos() - item.queue_enter_time);
+        closure->Run();
+        _blocked_closure_num[0]--;
+    }
+    _total_chunks--;
+    _recvr->_num_buffered_bytes -= item.chunk_bytes;
+    return true;
+}
+
+Status DataStreamRecvr::PipelineSenderQueue::add_chunks(const PTransmitChunkParams& request,
+                                                        ::google::protobuf::Closure** done) {
+    bool use_pass_through = request.use_pass_through();
+    DCHECK(request.chunks_size() > 0 || use_pass_through);
+    if (_is_cancelled || _num_remaining_senders <= 0) {
+        return Status::OK();
+    }
+
+    ScopedTimer<MonotonicStopWatch> wait_timer(_recvr->_sender_wait_lock_timer);
+    {
+        std::lock_guard<Mutex> l(_lock);
```

Comment:
Is it efficient to use a spinlock here?

Reply:
> Is it efficient to use a spinlock here?

I think it's ok, because we won't hold the lock for a long time